docs/contributing.md

```diff
@@ -2,11 +2,53 @@
 
 ## Testing
 
-Testing is done using "Vanilla" golang testing frameworks, with [require](https://godoc.org/github.com/stretchr/testify/require) for making assertions.
-Prefer to test a package's exported behaviour, rather than the implementation details.
+Testing is done using "vanilla" golang testing frameworks, with
+[Testify require](https://godoc.org/github.com/stretchr/testify/require) for making assertions. Require fails as soon as
+an assertion fails and so is preferred to [Testify assert](https://godoc.org/github.com/stretchr/testify/assert),
+although assert may be used where appropriate.
 
-### Test naming
+Nested `t.Run` assertions are preferred, as they lead to clear test names. For example:
 
-Test names should be of the form `$testObject_$testAction_$expectedResult`, for example:
-* `TestPeerController_OnCreation_CreatesReplicaSet`
-* `TestClusterController_WithInalidConfig_FailsToApply`
+```go
+func TestFoo(t *testing.T) {
+
+	t.Run("TestClusterController", func(t *testing.T) {
+		// Init Controller
+		t.Run("OnCreation", func(t *testing.T) {
+			//create cluster
+			t.Run("CreatesService", func(t *testing.T) {
+				//assert service exists
+				t.Fail()
+			})
+		})
+	})
+}
+```
+
+Would produce test output like the following:
+
+```
+--- FAIL: TestFoo/TestClusterController/OnCreation/CreatesService (0.00s)
+```
+
+### Unit Tests
+
+Unit tests should only be written for pure functions, and should not mock things like contexts or the Go Kubernetes
+client. Due to the nature of the project, unit tests are relatively rare, but encouraged for pure functions that do a
+lot of work or handle corner cases.
```

Comment:
as we've removed the comment above about preferring to test a package's exported behaviour, can we add something to the effect of "avoid using unit tests to validate the behaviour of package internals, unless the behaviour is sufficiently complex to warrant examples"